### PR TITLE
Do not retreive shell items via built-in COM

### LINKED ISSUE #6748: [dotnet-sdk-7.0.100-preview.3.22118.7] Greenshot and NugetPackageExplorerget apps get exception When do saving the file operation
### Description

When running Greenshot with .NET 7.0 Preview 3 runtime, it will get exception when do saving the file operation.

### Reproduction Steps

**Repro steps for the affected app:**

**This is .Net 5.0 app, but we are running it on 7.0, machine only have 7.0 installed, and DOTNET_ROLL_FORWARD_ON_NO_CANDIDATE_FX=2, so it runs on 7.0.**

1.Copy App from App Path and launch Greenshot.exe. 
2.Click “OK” button.
3.Press Print Screen key.
4.Take a screenshot by dragging the cursor down a little and releasing the left button click.
5.Choose Save as (displaying dialog) menu item.

**Minimal Repro steps:**
**dotnet-sdk-7.0.100-preview.322118.7 sdk should be installed on the machine.** 
Sample code attached. 
[WinFormsApp1 (2).zip](https://github.com/dotnet/runtime/files/8122661/WinFormsApp1.2.zip)

1) Create .net 5.0 Winform app
2) Add SaveFileDialog and Button Component to form
3) In Button click event, write this :
```
 private void button1_Click(object sender, EventArgs e)
        {
             var dialog = new System.Windows.Forms.SaveFileDialog()
            {
               InitialDirectory = @"C:\Users\v-zhiyang\Documents"
            };
            try
            {
                var result = dialog.ShowDialog();
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }

        }
```
4) Publish the app and change the runtime.json file to use 7.0.0-preview.3.22116.6 for Microsoft.WindowsDesktop.App under publish folder.
```
{
  "runtimeOptions": {
    "tfm": "net5.0",
    "framework": {
      "name": "Microsoft.WindowsDesktop.App",
      "version": "7.0.0-preview.3.22116.6"
    }
  }
}
```
5) Start the project with F5.
6) Click "button1" Button.

### Expected behavior

It shows save window.

### Actual behavior

**Actual Result for the Affected App:**
It shows error and crash.
```
Application: Greenshot.exe
Microsoft.WindowsDesktop.App version: 7.0.0-preview.3.22116.6
Description: The process was terminated due to an unhandled exception.
Exception Info: System.InvalidCastException: 
System.NotImplementedException: ComWrappers for type System.__ComObject not implemented.
   at Interop.WinFormsComWrappers.ComputeVtables(Object obj, CreateComInterfaceFlags flags, Int32& count)
   at System.Runtime.InteropServices.ComWrappers.<TryGetOrCreateComInterfaceForObjectInternal>g____PInvoke__|7_0(ObjectHandleOnStack comWrappersImpl, Int64 wrapperId, ObjectHandleOnStack instance, CreateComInterfaceFlags flags, IntPtr* retValue)
   at System.Runtime.InteropServices.ComWrappers.TryGetOrCreateComInterfaceForObjectInternal(ObjectHandleOnStack comWrappersImpl, Int64 wrapperId, ObjectHandleOnStack instance, CreateComInterfaceFlags flags, IntPtr& retValue)
   at System.Runtime.InteropServices.ComWrappers.TryGetOrCreateComInterfaceForObjectInternal(ComWrappers impl, Object instance, CreateComInterfaceFlags flags, IntPtr& retValue)
   at System.Runtime.InteropServices.ComWrappers.GetOrCreateComInterfaceForObject(Object instance, CreateComInterfaceFlags flags)
   at Interop.WinFormsComWrappers.GetOrCreateComInterfaceForObject(Object obj)
   at Interop.WinFormsComWrappers.GetComPointer[T](T obj, Guid iid)
   at Interop.WinFormsComWrappers.FileDialogWrapper.Interop.Shell32.IFileDialog.SetDefaultFolder(IShellItem psi)
   at System.Windows.Forms.FileDialog.OnBeforeVistaDialog(IFileDialog dialog)
   at System.Windows.Forms.FileDialog.TryRunDialogVista(IntPtr hWndOwner, Boolean& returnValue)
   at System.Windows.Forms.FileDialog.RunDialog(IntPtr hWndOwner)
   at System.Windows.Forms.CommonDialog.ShowDialog(IWin32Window owner)
   at Greenshot.Addons.Controls.SaveImageFileDialog.ShowDialog() in C:\.NETCoreUpgradeApps\AppSources\greenshot\src\Greenshot.Addons\Controls\SaveImageFileDialog.cs:line 212
   at Greenshot.Addons.Core.ImageOutput.SaveWithDialog(ISurface surface, ICaptureDetails captureDetails) in C:\.NETCoreUpgradeApps\AppSources\greenshot\src\Greenshot.Addons\Core\ImageOutput.cs:line 122
   at Greenshot.Destinations.FileWithDialogDestination.ExportCapture(Boolean manuallyInitiated, ISurface surface, ICaptureDetails captureDetails) in C:\.NETCoreUpgradeApps\AppSources\greenshot\src\Greenshot\Destinations\FileWithDialogDestination.cs:line 59
   at Greenshot.Addons.Core.AbstractDestination.ExportCaptureAsync(Boolean manuallyInitiated, ISurface surface, ICaptureDetails captureDetails) in C:\.NETCoreUpgradeApps\AppSources\greenshot\src\Greenshot.Addons\Core\AbstractDestination.cs:line 159
   at Greenshot.Addons.Core.AbstractDestination.<>c__DisplayClass42_0.<<ShowPickerMenu>b__6>d.MoveNext() in C:\.NETCoreUpgradeApps\AppSources\greenshot\src\Greenshot.Addons\Core\AbstractDestination.cs:line 367
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

```

**Actual Result for the Minimal Demo:**

Get below exception:
System.NotImplementedException: ComWrappers for type System.__ComObject not implemented.

### Regression?

Yes

**Verify Scenarios:**

1). Windows10 RS5 x64+dotnet-sdk-7.0.100-preview.3.22118.7: Fail
2). Windows10 RS5 x64+dotnet-sdk-7.0.100-preview.3.22122.1: Fail
3). Windows10 RS5 x64+dotnet-sdk-7.0.100-preview.2.22114.1: Pass
4). Windows10 RS5 x64+dotnet-sdk-5.0.405-win-x64 (Default SDK): Pass

### Known Workarounds

_No response_

### Configuration

**Application Name**: Greenshot  & NugetPackageExplorer
**OS**: Windows 10 RS5
**CPU**: X64
**.NET Build Number:** dotnet-sdk-7.0.100-preview.3.22118.7

**Github link for source code**:
https://github.com/greenshot/greenshot 

**App and App Source Checking at** : https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1487350

### Other information

**Findings:**

When we create SaveFileDialog object, InitialDirectory will report an exception：
       InitialDirectory = @"C:\Users\v-zhiyang\Documents"

@dotnet-actwx-bot @dotnet/compat

### PR DESCRIPTION
**Pull request by kant2002**
Fixed #6748

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6751)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateItemFromParsingName.cs [hunk 1]
```diff
@@ -2,13 +2,27 @@
...
+        public static extern HRESULT SHCreateItemFromParsingName(string pszPath, IntPtr pbc, in Guid riid, out IntPtr ppv);
+
+        public static IShellItem CreateItemFromParsingName(string path)
+        {
+            Guid shellItemIID = IID.IShellItem;
+            HRESULT hr = SHCreateItemFromParsingName(path, IntPtr.Zero, in shellItemIID, out IntPtr ppv);
+            if (hr.Failed())
+            {
+                throw new Win32Exception((int)hr);
+            }
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateItemFromParsingName.cs:22]: Can we use `hr.ThrowIfFailed()`?

**Comment by RussKie** [comment]
Standing by for a confirmation that you tested it locally.

File: src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateItemFromParsingName.cs (hunk 1 above)
Comment (kant2002) [reply src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateItemFromParsingName.cs:22]: Wouldn’t this be breaking change? It’s not that this is super important, but who knows

**Comment by kant2002** [comment]
Is there a way how I can get list of versions for daily installers? I do want to go back in time and find working SDK for me.

**Comment by RussKie** [comment]
> Is there a way how I can get list of versions for daily installers? I do want to go back in time and find working SDK for me.

@mmitche @marcpopMSFT are you able to help? Do we have a list of daily installer builds available somewhere publicly?

File: src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateItemFromParsingName.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/src/Interop/Shell32/Interop.SHCreateItemFromParsingName.cs:22]: Oh, you moved it from FolderBrowserDialog. Let's keep it then.

**Comment by kant2002** [comment]
@RussKie what version do you working with. I would like to know at least some version which is working, since either I have mess up something locally or there some issue which I cannot diagnose.

**Comment by RussKie** [comment]
> what version do you working with. I would like to know at least some version which is working, since either I have mess up something locally or there some issue which I cannot diagnose.

How are you testing? I checked out your branch, and was able run WinformsControlsTest project from VS.

**Comment by kant2002** [comment]
I test using supplied in the issue app using globally installed version of .NET SDK 7.0.100-preview.1.22114.7 and daily build.
I simply run using dotnet run.
In order to test changes, I do modify csproj to supply project with localy build WinForms.

Will try in a way how do you do testing.

**Comment by kant2002** [comment]
@RussKie I did test using WinformsControlsTest and setting InitialDirectory did not trigger the issue anymore.

**Comment by mmitche** [comment]
> > Is there a way how I can get list of versions for daily installers? I do want to go back in time and find working SDK for me.
> 
> @mmitche @marcpopMSFT are you able to help? Do we have a list of daily installer builds available somewhere publicly?

Not really, but sorta. You could walk through a list of builds on main, but that's only available to internal folks. But, for public builds, you can use a bit of a shortcut. Every time the installer builds, it pushes a package called Microsoft.DotNet.Sdk.Internal which we use for dependency flow. The version of this package corresponds to an available SDK version:

https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet7-transport/NuGet/Microsoft.Dotnet.Sdk.Internal/7.0.100-preview.3.22123.26/versions

Note that due to an AzDO rollout issue, this link may not currently be visible to external folks (I'm told this is being fixed). But a nuget browser like VS should be able to surface the same information against the dotnet7-transport nuget feed: https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet7-transport/nuget/v3/index.json

**Comment by RussKie** [comment]
> I test using supplied in the issue app using globally installed version of .NET SDK 7.0.100-preview.1.22114.7 and daily build. I simply run using dotnet run. In order to test changes, I do modify csproj to supply project with localy build WinForms.
> 
> Will try in a way how do you do testing.

I will be updating the https://github.com/dotnet/winforms/blob/main/docs/debugging.md guide explaining the steps. Thank you.

**Comment by kant2002** [comment]
Would like to have this land master, so next Preview would not have this issue. This is stabilizes my File Dialog work.

**Review by RussKie — APPROVED** [review]

**Comment by kant2002** [comment]
@RussKie I did small app which walk `Version.Details.xml` and look for dependencies at specific point of time. https://kant2002.github.io/dotnet-installer-viewer/

Can you or @mmitche validate that I capture existing process properly? I hope that allows me faster find what version of runtime and winforms are bundled in the specific installer.

**Comment by RussKie** [comment]
This is awesome! 💯 

Here's your tool's output:
image

Here's our internal tool's output:
image

**Pull request merged**